File: C#/Easy/ValidParentheses.cs
public class Solution
{
    public bool IsValid(string s)
    {
        var stack = new Stack<char>();

        foreach (char c in s)
        {
            if (c == '{' || c == '[' || c == '(')
            {
                stack.Push(c);
            }
            else
            {
                if (stack.Count == 0) return false;
                
                char c1 = stack.Pop();
                if (c == '}' && c1 != '{') return false;
                if (c == ']' && c1 != '[') return false;
                if (c == ')' && c1 != '(') return false;
            }
        }

        return stack.Count == 0;
    }
}

/*
    
    Time Complexity: O(n)
    Space Complexity: O(n)
    
    Where n is the length of the string

    Key insights: 
    * Once you validate a opening/closing pair, you can ignore it
    * You're always comparing with the most recent open bracket

*/

File: Swift/Easy/ValidParentheses.swift
class Solution {
    func isValid(_ s: String) -> Bool {
        var stack = [Character]()
        for char in s {
            if char == "(" || char == "[" || char == "{" {
                stack.append(char);
            }
            else {
                if stack.isEmpty { return false }
                let last = stack.removeLast()
                if (char == ")" && last != "(") || 
                    (char == "]" && last != "[") || 
                    (char == "}" && last != "{") { 
                    return false
                }
            }
        }

        return stack.isEmpty
    }
}

/*

    Time: O(n)
    Space: O(n)

    Where n is the length of the input string

*/



class Solution2 {
    func isValid(_ s: String) -> Bool {
        var map = [Character: Character]()
        var stack = [Character]()
        map["("] = ")"
        map["["] = "]"
        map["{"] = "}"

        for char in s {
            if let temp = map[char] {
                stack.append(temp)
                continue
            }

            if stack.isEmpty { return false }
            let last = stack.removeLast()
            if last != char { return false }
        }

        return stack.isEmpty
    }
}

/*

    Time: O(n)
    Space: O(n)

    Same solution as the first but the mappings are stored in a dictionary

*/
